modules/assim.sequential/R/Analysis_sda_block.R

```diff
@@ -0,0 +1,549 @@
+        block.list[[i]]$Inits$q <- c(block.list[[i]]$Inits$q, stats::rgamma(1, shape = block.list[[i]]$data$aq[j], rate = block.list[[i]]$data$bq[j]))
+      }
+    } else if (block.list[[i]]$constant$q.type == 2) {
+      #if we want the wishart Q.
+      if ("try-error" %in% class(try(block.list[[i]]$Inits$q <- 
+                                     stats::rWishart(1, df = block.list[[i]]$data$bq, Sigma = block.list[[i]]$data$aq)[,,1], silent = T))) {
+        block.list[[i]]$Inits$q <- 
+          stats::rWishart(1, df = block.list[[i]]$data$bq, Sigma = stats::toeplitz((block.list[[i]]$constant$YN:1)/block.list[[i]]$constant$YN))[,,1]
+      }
+    }
+  }
+  #return values.
+  return(block.list)
+}
+
+##' @title MCMC_block_function
+##' @name  MCMC_block_function
+##' @author Dongchen Zhang
+##' 
+##' @param block  each block within the `block.list` lists.
+##' 
+##' @return It returns the `block` object with analysis results filled in.
+MCMC_block_function <- function(block) {
+  #build nimble model
+  model_pred <- nimble::nimbleModel(GEF.Block.Nimble,
+                                    data = block$data,
+                                    inits = block$Inits,
+                                    constants = block$constant,
+                                    name = 'base')
+  #configure MCMC
+  conf <- nimble::configureMCMC(model_pred, print=FALSE)
+  conf$setMonitors(c("X", "X.mod", "q"))
+  
+  #Handle samplers
+  #hear we change the RW_block sampler to the ess sampler 
+  #because it has a better performance of MVN sampling
+  samplerLists <- conf$getSamplers()
+  samplerNumberOffset <- length(samplerLists)
+  if (block$constant$q.type == 1) {
+    #if we have vector q
+    #only X.mod should be sampled with ess sampler.
+    X.mod.ind <- which(grepl("X.mod", samplerLists %>% purrr::map(~ .x$target) %>% unlist()))
+    samplerLists[[X.mod.ind]]$setName("ess")
+  } else if (block$constant$q.type == 2) {
+    #if we have wishart q
+    #everything should be sampled with ess sampler.
+    samplerLists %>% purrr::map(function(l){l$setName("ess")})
+  }
+  conf$setSamplers(samplerLists)
+  
+  #add toggle Y sampler.
+  for (i in 1:block$constant$YN) {
+    conf$addSampler(paste0("y.censored[", i, "]"), 'toggle', control=list(type='RW'))
+  }
+  conf$printSamplers()
+  #compile MCMC
+  Rmcmc <- nimble::buildMCMC(conf)
+  Cmodel <- nimble::compileNimble(model_pred)
+  Cmcmc <- nimble::compileNimble(Rmcmc, project = model_pred, showCompilerOutput = FALSE)
+  
+  #if we don't have any NA in the Y.
+  if (!any(is.na(block$data$y.censored))) {
+    #add toggle Y sampler.
+    for(i in 1:block$constant$YN) {
+      valueInCompiledNimbleFunction(Cmcmc$samplerFunctions[[samplerNumberOffset+i]], 'toggle', 0)
+    }
+  }
+  
+  #run MCMC
+  dat <- runMCMC(Cmcmc, niter = block$MCMC$niter, nburnin = block$MCMC$nburnin, thin = block$MCMC$nthin, nchains = block$MCMC$nchain)
+  
+  #update aq, bq, mua, and pa
```

Comment:
Added a note: `#TODO: harmonize the MCMC code between block-based and general analysis functions to reduce the code complexity.` inside the `MCMC_block_function`
This will be a future PR sometime.